### DISCUSSION TITLE
[CRUD Service] Introspection Endpoint

### DISCUSSION THREAD

**Discussion by GioAc96** [mia-platform/community#350] [Ideas]
### Which area of this project could be most improved?

Marketplace

### Description

I would like to propose a new feature for the CRUD Service: collection introspection. I believe it would be nice if the CRUD service exposed an endpoint that returned the JSON schema of each collection, as shown below:

Request:
```
GET http://crud-service/books/introspect
```
Response:
```json
{
  "schema": {
    "type": "object",
    "properties": {...},
  }
}
```

It would also be nice if the CRUD service had a generic introspection endpoint that allowed to discover all available collections, such as this:

Request:
```
GET http://crud-service/$introspect
```
Response:
```json
{
  "version": "1.2.3",
  "collections": {
    "books": {
      "schema": {...}
    }
  }
}
```

# Use Case
I believe this would benefit CRUD Service users by simplifying access to documentation about collections, without the need of API Portals and Swaggers. It would also be useful to be able to easily download JSON Schemas for collections and integrate such schemas with code generating tools (e.g: automatically generating data Classes in OO languages), therefore simplifying the end user experience.

**Comment by fredmaggiowski** [comment]
The CRUD already exposes the OpenAPI Specification, isn't that enough to find out what collections are available and their data model? 🤔

**Comment by danibix95** [comment]
Hi @GioAc96, the version `v6.10.0` of CRUD Service should match the proposed idea. Indeed, it introduce

- a new endpoint `<collection>/schema` for each collection, which allows to retrieve the collection data model as a JSON Schema
- a new general endpoint (with configurable prefix) `/-/schemas` for retrieving the JSON Schema of all the collections associated to the specific CRUD Service instance. The response can be either an `application/json` or `application/x-ndjson`

**Reply by GioAc96** [reply]
Awesome, thank you!
